File: src/controller/chore.controller.ts
import {
  Controller,
  Get,
  Post,
  Patch,
  Delete,
  Param,
  Query,
  Body,
  HttpCode,
  HttpStatus,
  BadRequestException,
} from '@nestjs/common';
import { ChoreService } from '../service/chore.service';
import { NoDescriptionChoreDTO } from './dto/outgoing/noDescriptionChore.dto';
import { NoIdChoreDTO } from './dto/outgoing/noIdChore.dto';
import { ChoreDTO } from './dto/outgoing/chore.dto';
import { NewChoreDTO } from './dto/incoming/newChore.dto';
import { UpdateChoreDTO } from './dto/incoming/updateChore.dto';

@Controller('chores')
export class ChoreController {
  constructor(private readonly choreService: ChoreService) {}

  @Get()
  async findAll(
    @Query('done') done?: string
  ): Promise<NoDescriptionChoreDTO[]> {
    return this.choreService.findAll(done ? done === 'true' : undefined);
  }

  @Get(':id')
  async findById(@Param('id') id: number): Promise<NoIdChoreDTO> {
    if (!id) {
      throw new BadRequestException('Id must be a number');
    }
    return this.choreService.findById(id);
  }

  @Post()
  @HttpCode(HttpStatus.CREATED)
  async create(@Body() choreData: NewChoreDTO): Promise<ChoreDTO> {
    const id = await this.choreService.create({ ...choreData, done: false });
    return { ...choreData, id, done: false } as ChoreDTO;
  }

  @Patch(':id')
  @HttpCode(HttpStatus.NO_CONTENT)
  async update(
    @Param('id') id: number,
    @Body() choreData: UpdateChoreDTO
  ): Promise<void> {
    if (!id) {
      throw new BadRequestException('Id must be a number');
    }
    await this.choreService.update(id, choreData);
  }

  @Delete(':id')
  @HttpCode(HttpStatus.NO_CONTENT)
  async delete(@Param('id') id: number): Promise<void> {
    if (!id) {
      throw new BadRequestException('Id must be a number');
    }
    await this.choreService.delete(id);
  }
}

File: src/controller/chore_controller.py
from flask import Blueprint, request, jsonify, abort
from marshmallow import ValidationError

from src.controller.dto.incoming_dto import new_chore_schema, update_chore_schema
from src.controller.dto.outgoing_dto import no_id_chore_schema, chore_schema, no_description_chore_schema
from src.service.chore_service import get_all_chores, get_chore_by_id, create_chore, update_chore, delete_chore
from src.service.exceptions import ChoreNotFoundException

chore_bp = Blueprint('chore', __name__)


@chore_bp.route('', methods=['GET'])
def get_chores():
    return jsonify(no_description_chore_schema.dump(get_all_chores(request.args.get('done')), many=True))


@chore_bp.route('/<int:chore_id>', methods=['GET'])
def get_chore(chore_id):
    try:
        chore = get_chore_by_id(chore_id)
        return jsonify(no_id_chore_schema.dump(chore))
    except ChoreNotFoundException:
        abort(404, 'Chore not found')


@chore_bp.route('', methods=['POST'])
def create():
    try:
        chore_data = new_chore_schema.load(request.json)
        return jsonify(chore_schema.dump(create_chore(chore_data))), 201
    except ValidationError as err:
        abort(400, err.messages)


@chore_bp.route('/<int:chore_id>', methods=['PATCH'])
def update(chore_id):
    try:
        chore_data = update_chore_schema.load(request.json)
        update_chore(chore_id, chore_data)
        return '', 204
    except ValidationError as err:
        abort(400, err.messages)
    except ChoreNotFoundException:
        abort(404, 'Chore not found')


@chore_bp.route('/<int:chore_id>', methods=['DELETE'])
def delete(chore_id):
    delete_chore(chore_id)
    return '', 200
